src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToNewComponentCodeActionProvider.cs

```diff
@@ -47,51 +51,166 @@ public Task<ImmutableArray<RazorVSInternalCodeAction>> ProvideAsync(RazorCodeAct
+    /// <param name="actionParams">The parameters for the extraction action, which will be updated.</param>
+    private static void ProcessMultiPointSelection(MarkupElementSyntax startElementNode, MarkupElementSyntax? endElementNode, ExtractToNewComponentCodeActionParams actionParams)
+    {
+        // If there's no end element, we can't process a multi-point selection
+        if (endElementNode is null)
         {
-            Action = LanguageServerConstants.CodeActions.ExtractToNewComponentAction,
-            Language = LanguageServerConstants.CodeActions.Languages.Razor,
-            Data = actionParams,
-        };
+            return;
+        }
 
-        var codeAction = RazorCodeActionFactory.CreateExtractToNewComponent(resolutionParams);
+        // Check if the start element is an ancestor of the end element
+        var selectionStartHasParentElement = endElementNode.Ancestors().Any(node => node == startElementNode);
 
-        return Task.FromResult<ImmutableArray<RazorVSInternalCodeAction>>([codeAction]);
+        // If the start element is an ancestor, keep the original end; otherwise, use the end of the end element
+        actionParams.ExtractEnd = selectionStartHasParentElement ? actionParams.ExtractEnd : endElementNode.Span.End;
+
+        // If the start element is not an ancestor of the end element, we need to find a common parent
+        // This conditional handles cases where the user's selection spans across different levels of the DOM.
+        // For example:
+        //   <div>
+        //     <span>
+        //      Selected text starts here<p>Some text</p>
+        //     </span>
+        //     <span>
+        //       <p>More text</p>
+        //     </span>
+        //     Selected text ends here <span></span>
+        //   </div>
+        // In this case, we need to find the smallest set of complete elements that covers the entire selection.
+        if (!selectionStartHasParentElement)
+        {
+            // Find the closest containing sibling pair that encompasses both the start and end elements
+            var (extractStart, extractEnd) = FindContainingSiblingPair(startElementNode, endElementNode);
+
+            // If we found a valid containing pair, update the extraction range
+            if (extractStart is not null && extractEnd is not null)
+            {
+                actionParams.ExtractStart = extractStart.Span.Start;
+                actionParams.ExtractEnd = extractEnd.Span.End;
+            }
+            // Note: If we don't find a valid pair, we keep the original extraction range
+        }
+    }
+
+    private static bool IsMultiPointSelection(Range range) => range.Start != range.End;
+
+    private static SourceLocation? GetEndLocation(Position selectionEnd, RazorCodeDocument codeDocument, ILogger logger)
+    {
+        if (!selectionEnd.TryGetSourceLocation(codeDocument.GetSourceText(), logger, out var location))
```

Comment:
Note: `GetSourceText()` has been removed in main. You can avoid that merge conflict by changing this to `codeDocument.Source.Text`.